faiss/utils/hamming-inl.h

```diff
@@ -226,19 +227,52 @@ struct HammingComputerDefault {
     }
 
     void set (const uint8_t *a8, int code_size) {
-        a =  a8;
-        n = code_size;
+        this->a8 = a8;
+        quotient8 = code_size / 8;
+        remainder8 = code_size % 8;
     }
 
     int hamming (const uint8_t *b8) const {
         int accu = 0;
-        for (int i = 0; i < n; i++)
-            accu += popcount64 (a[i] ^ b8[i]);
+
+        auto a64 = reinterpret_cast<const uint64_t *>(a8);
+        auto b64 = reinterpret_cast<const uint64_t *>(b8);
+        int i = 0, len = quotient8;
+        switch (len & 7) {
+            default:
+                while (len > 7) {
+                    len -= 8;
+                    accu += popcount64(a64[i] ^ b64[i]); i++;
+                    case 7: accu += popcount64(a64[i] ^ b64[i]); i++;
+                    case 6: accu += popcount64(a64[i] ^ b64[i]); i++;
+                    case 5: accu += popcount64(a64[i] ^ b64[i]); i++;
+                    case 4: accu += popcount64(a64[i] ^ b64[i]); i++;
+                    case 3: accu += popcount64(a64[i] ^ b64[i]); i++;
+                    case 2: accu += popcount64(a64[i] ^ b64[i]); i++;
+                    case 1: accu += popcount64(a64[i] ^ b64[i]); i++;
+                }
+        }
+        if (remainder8) {
+            auto a = a8 + 8 * quotient8;
+            auto b = b8 + 8 * quotient8;
+            switch (remainder8) {
+                case 7: accu += hamdis_tab_ham_bytes[a[6] ^ b[6]];
```

Comment:
I suspect the table look-up is slower than using popcount64

Reply:
Here we calculate 1 to 7 uint8.
And my local tests show that table look-up is much faster than using popcount64.

benchmark with code_size=127 n=4000000 nrun=100
Hamming_Dft_lookup  implem: 550.326 ms
Hamming_Dft_popcnt  implem: 743.236 ms
benchmark with code_size=255 n=4000000 nrun=100
Hamming_Dft_lookup  implem: 1117.139 ms
Hamming_Dft_popcnt  implem: 1262.516 ms